stor/dx.py

```diff
@@ -0,0 +1,513 @@
+            # first part can be 'proj:file' or 'proj:' or 'proj'
+            parts_first = parts[0].split(':')
+            parts[0] = parts_first[0]
+            if len(parts_first) > 1 and parts_first[1]:
+                parts.insert(1, parts_first[1])
+        return parts
+
+    def _is_folder(self):
+        return self.resource and not self.resource.ext and \
+               not utils.is_valid_dxid(self.resource.lstrip('/'), 'file')
+
+    def _noop(attr_name):
+        def wrapper(self):
+            return type(self)(self)
+        wrapper.__name__ = attr_name
+        wrapper.__doc__ = 'No-op for %r' % attr_name
+        return wrapper
+
+    abspath = _noop('abspath')
+    realpath = _noop('realpath')
+    expanduser = _noop('expanduser')
+
+    @property
+    def project(self):
+        """Returns the project name from the path or None"""
+        parts = self._get_parts()
+        return parts[0] if len(parts) > 0 and parts[0] else None
+
+    def temp_url(self, lifetime=300, filename=None):
+        """Obtains a temporary URL to a DNAnexus data-object.
+
+        Args:
+            lifetime (int): The time (in seconds) the temporary
+                URL will be valid
+            filename (str, optional): A urlencoded filename to use for
+                attachment, otherwise defaults to object name
+        """
+        try:
+            if self.canonical_resource:
+                file_handler = dxpy.DXFile(self.canonical_resource)
+                return file_handler.get_download_url(
+                    duration=lifetime,
+                    preauthenticated=True,
+                    filename=filename,
+                    project=self.canonical_project
+                )[0]
+            else:
+                raise DXError('DX Projects cannot have a temporary download url')
+        except ValueError:
+            raise DXError('DXPaths ending in folders cannot have a temporary download url')
+
+    @property
+    def resource(self):
+        parts = self._get_parts()
+        joined_resource = '/'.join(parts[1:]) if len(parts) > 1 else None
+        return self.parts_class('/'+joined_resource) if joined_resource else None
+
+    def dirname(self):
+        if not self.resource:
+            return self
+        else:
+            return super(DXPath, self).dirname()
+
+    def download_objects(self):  # may not need it
+        raise NotImplementedError
+
+    def remove(self):
+        raise NotImplementedError
+
+    def rmtree(self):
+        raise NotImplementedError
+
+    def isdir(self):
+        if not self.resource or self._is_folder():
+            return self.exists()
+        return False
+
+    def isfile(self):
+        try:
+            return self.resource and not self._is_folder() and self.exists()
+        except NotFoundError:
+            return False
+
+    def getsize(self):
+        if not self.resource:
+            return self.stat()['dataUsage']*1e9
+        elif self._is_folder():
```

Comment:
is the `_is_folder` check necessary here? `getsize()` returns something nonsensical for folders on PosixPaths and raises a `NotFoundError` for swift and s3 paths. To me seems like you can just get rid of this clause completely.

Also this should have a docstring that describes what it can raise and what it returns.